### PR TITLE
Ensure that 'extra_params' get applied to template-related objects (ZenPack)

### PR DESCRIPTION
**Pull request by j053ph4**
Fixes ZEN-24083

- update 'apply_data_defaults' to explicitly update 'extra_params' from DEFAULTS if not explicitly set
- updated ZenPackSpec to allow examination of template objects in memory
- added methods to TestHarness to compare tempate-related objects to specs and yaml
- updated unit tests to validate templates vs. specs and vs. yaml
- removed unused code from test_classes.py
- change ZPLCommand default connect parameter to False

### CODE REVIEW

File: ZenPacks/zenoss/ZenPackLib/tests/ZPLTestHarness.py [hunk 1]
```diff
@@ -268,6 +278,116 @@ def compare_ob_to_spec(self, ob):
...
+                            passed = False
+        return passed
+
+    def compare_template_to_spec(self, template, spec):
+        '''compare template to spec'''
+        if template.id != spec.name:
+            return False
+        if template.description != spec.description:
+            return False
+        if tcs.targetPythonClass:
```
Comment (pcarinhas) [inline ZenPacks/zenoss/ZenPackLib/tests/ZPLTestHarness.py]: (tcs, t) undefined in function

File: ZenPacks/zenoss/ZenPackLib/tests/ZPLTestHarness.py [hunk 2]
```diff
@@ -268,6 +278,116 @@ def compare_ob_to_spec(self, ob):
                 passed = False
         return passed
```
Comment (pcarinhas) [inline ZenPacks/zenoss/ZenPackLib/tests/ZPLTestHarness.py]: In Line 135, d is undefined.

File: ZenPacks/zenoss/ZenPackLib/tests/ZPLTestHarness.py (hunk 1 above)
Comment (j053ph4) [inline ZenPacks/zenoss/ZenPackLib/tests/ZPLTestHarness.py]: good catch i fixed it

File: ZenPacks/zenoss/ZenPackLib/tests/ZPLTestHarness.py [hunk 3]
```diff
@@ -1,17 +1,27 @@
 import inspect
```
Comment (pcarinhas) [inline ZenPacks/zenoss/ZenPackLib/tests/ZPLTestHarness.py]: ZPLTestHarness.py|1 col 1 error| 'inspect' imported but unused [F401]
ZPLTestHarness.py|165 col 9 error| local variable 'bases_from' is assigned to but never used [F841]
ZPLTestHarness.py|166 col 9 error| local variable 'bases_to' is assigned to but never used [F841]

File: ZenPacks/zenoss/ZenPackLib/lib/spec/Spec.py [hunk 4]
```diff
@@ -185,6 +185,12 @@ def apply_data_defaults(self, dictionary, default_defaults=None, leave_defaults=
                 defaults = dictionary.pop('DEFAULTS')
             for k, v in dictionary.iteritems():
                 dictionary[k] = dict(defaults, **v)
+                if 'extra_params' in  dictionary[k].keys():
```
Comment (pcarinhas) [inline ZenPacks/zenoss/ZenPackLib/lib/spec/Spec.py]: Spec.py|188 col 37 error| multiple spaces after keyword [E271]

File: ZenPacks/zenoss/ZenPackLib/tests/ZPLTestHarness.py (hunk 2 above)
Comment (j053ph4) [inline ZenPacks/zenoss/ZenPackLib/tests/ZPLTestHarness.py]: fixed

**Pull request merged**